File: src/bmsx/render/presentation_state.ts
import { consoleCore } from '../core/console';
import type { Runtime } from '../machine/runtime/runtime';
import type { TickCompletion } from '../machine/scheduler/frame';
import * as workbenchMode from '../ide/workbench/mode';
import { commitVdpViewSnapshot } from './vdp/view_snapshot';

export type RenderPresentationMode = 'partial' | 'completed';

type RenderPresentation = {
	mode: RenderPresentationMode;
	commitFrame: boolean;
};

export class RenderPresentationState {
	private pendingPresentation = false;
	private presentationMode: RenderPresentationMode = 'completed';
	private presentationCommitFrame = false;
	private debugPresentReportAtMs = 0;
	private debugPresentHostFrames = 0;
	private debugPresentTickCompleted = 0;
	private debugPresentTickCommitted = 0;
	private debugPresentTickDeferred = 0;
	private debugPresentTickHeld = 0;
	private debugPresentPartialPresents = 0;
	private debugPresentCommitPresents = 0;
	private debugPresentHoldPresents = 0;
	private debugPresentPausedPresents = 0;
	private readonly presentationScratch: RenderPresentation = {
		mode: 'completed',
		commitFrame: false,
	};
	private readonly tickCompletionScratch: TickCompletion = {
		sequence: 0,
		remaining: 0,
		visualCommitted: true,
		vdpFrameCost: 0,
		vdpFrameHeld: false,
	};

	constructor(private readonly runtime: Runtime) {
	}

	private recordTickCompletion(visualCommitted: boolean, vdpFrameHeld: boolean): void {
		if (!Boolean((globalThis as any).__bmsx_debug_presentrate)) {
			return;
		}
		this.debugPresentTickCompleted += 1;
		if (visualCommitted) {
			this.debugPresentTickCommitted += 1;
		} else {
			this.debugPresentTickDeferred += 1;
		}
		if (vdpFrameHeld) {
			this.debugPresentTickHeld += 1;
		}
	}

	private recordPresentation(mode: RenderPresentationMode, commitFrame: boolean): void {
		if (!Boolean((globalThis as any).__bmsx_debug_presentrate)) {
			return;
		}
		if (consoleCore.paused) {
			this.debugPresentPausedPresents += 1;
			return;
		}
		if (mode === 'partial') {
			this.debugPresentPartialPresents += 1;
			return;
		}
		if (commitFrame) {
			this.debugPresentCommitPresents += 1;
			return;
		}
		this.debugPresentHoldPresents += 1;
	}

	private resetDebugCounters(reportAtMs: number): void {
		this.debugPresentReportAtMs = reportAtMs;
		this.debugPresentHostFrames = 0;
		this.debugPresentTickCompleted = 0;
		this.debugPresentTickCommitted = 0;
		this.debugPresentTickDeferred = 0;
		this.debugPresentTickHeld = 0;
		this.debugPresentPartialPresents = 0;
		this.debugPresentCommitPresents = 0;
		this.debugPresentHoldPresents = 0;
		this.debugPresentPausedPresents = 0;
	}

	private presentFrame(hostDeltaMs: number, mode: RenderPresentationMode, commitFrame = mode === 'completed'): void {
		const runtime = this.runtime;
		consoleCore.deltatime = hostDeltaMs;
		runtime.machine.vdp.drainFrameBufferPresentation(consoleCore.view.vdpFrameBufferTextures);
		runtime.machine.vdp.drainSurfaceUploads(consoleCore.view.vdpSlotTextures);
		commitVdpViewSnapshot(consoleCore.view, runtime.machine.vdp.readDeviceOutput());
		consoleCore.view.configurePresentation(mode, commitFrame);
		this.recordPresentation(mode, commitFrame);
		consoleCore.sndmaster.finishFrame();
		consoleCore.view.drawgame();
	}

	private markPresentation(mode: RenderPresentationMode, commitFrame: boolean): void {
		this.pendingPresentation = true;
		this.presentationMode = mode;
		this.presentationCommitFrame = commitFrame;
	}

	public requestHeldPresentation(): void {
		if (!this.pendingPresentation) {
			this.markPresentation('completed', false);
		}
	}

	private consumePresentation(out: RenderPresentation): boolean {
		if (!this.pendingPresentation) {
			return false;
		}
		const runtime = this.runtime;
		const overlayActive = runtime.executionOverlayActive;
		out.mode = this.presentationMode;
		out.commitFrame = overlayActive ? false : this.presentationCommitFrame;
		workbenchMode.tickIDEDraw(runtime);
		workbenchMode.tickTerminalModeDraw(runtime);
		this.clearPresentation();
		return true;
	}

	public beginHostFrame(currentTime: number): void {
		if (!Boolean((globalThis as any).__bmsx_debug_presentrate)) {
			return;
		}
		if (this.debugPresentReportAtMs === 0) {
			this.debugPresentReportAtMs = currentTime;
		}
		this.debugPresentHostFrames += 1;
	}

	public clearPresentation(): void {
		this.pendingPresentation = false;
		this.presentationMode = 'completed';
		this.presentationCommitFrame = false;
	}

	public reset(): void {
		this.clearPresentation();
		this.resetDebugCounters(0);
	}

	public runOverlay(): void {
		const runtime = this.runtime;
		this.clearPresentation();
		if (runtime.frameLoop.currentFrameState !== null) {
			runtime.frameLoop.abandonFrameState();
		}
		runtime.frameScheduler.clearQueuedTime();
		workbenchMode.tickIDE(runtime);
		workbenchMode.tickTerminalMode(runtime);
		this.markPresentation('completed', false);
	}

	public syncAfterRuntimeUpdate(previousTickSequence: number): void {
		const runtime = this.runtime;
		if (runtime.executionOverlayActive) {
			runtime.frameScheduler.clearQueuedTime();
			this.markPresentation('completed', false);
		} else if (runtime.frameScheduler.lastTickSequence !== previousTickSequence) {
			this.markPresentation('completed', runtime.frameScheduler.lastTickVisualFrameCommitted);
		} else if (runtime.isDrawPending || runtime.workbenchFaultState.faultSnapshot !== null) {
			this.markPresentation('partial', false);
		}
		while (runtime.frameScheduler.consumeTickCompletion(this.tickCompletionScratch)) {
			this.recordTickCompletion(this.tickCompletionScratch.visualCommitted, this.tickCompletionScratch.vdpFrameHeld);
		}
	}


	public presentPausedFrame(hostDeltaMs: number): void {
		const runtime = this.runtime;
		if (runtime.executionOverlayActive) {
			this.runOverlay();
			this.consumePresentation(this.presentationScratch);
			this.presentFrame(hostDeltaMs, this.presentationScratch.mode, this.presentationScratch.commitFrame);
			return;
		}
		runtime.frameScheduler.clearQueuedTime();
		this.clearPresentation();
		this.presentFrame(hostDeltaMs, 'completed', false);
	}

	public presentPending(hostDeltaMs: number): boolean {
		if (!this.consumePresentation(this.presentationScratch)) {
			return false;
		}
		this.presentFrame(hostDeltaMs, this.presentationScratch.mode, this.presentationScratch.commitFrame);
		return true;
	}

	public presentErrorOverlay(hostDeltaMs: number): void {
		const runtime = this.runtime;
		if (!runtime.executionOverlayActive) {
			return;
		}
		this.runOverlay();
		this.consumePresentation(this.presentationScratch);
		this.presentFrame(hostDeltaMs, this.presentationScratch.mode, this.presentationScratch.commitFrame);
	}

	public flushDebugReport(currentTime: number, runtime: Runtime): void {
		if (!Boolean((globalThis as any).__bmsx_debug_presentrate)) {
			return;
		}
		if (this.debugPresentReportAtMs === 0) {
			this.debugPresentReportAtMs = currentTime;
			return;
		}
		const elapsedMs = currentTime - this.debugPresentReportAtMs;
		if (elapsedMs < 1000) {
			return;
		}
		const scale = 1000 / elapsedMs;
		const hostFps = this.debugPresentHostFrames * scale;
		console.warn(
			`[BMSX][present] host_frames=${this.debugPresentHostFrames} host_fps=${hostFps.toFixed(2)} ufps=${runtime.timing.ufps.toFixed(2)} `
			+ `tick_completed=${this.debugPresentTickCompleted} tick_committed=${this.debugPresentTickCommitted} `
			+ `tick_deferred=${this.debugPresentTickDeferred} tick_held=${this.debugPresentTickHeld} `
			+ `present_partial=${this.debugPresentPartialPresents} present_commit=${this.debugPresentCommitPresents} `
			+ `present_hold=${this.debugPresentHoldPresents} present_paused=${this.debugPresentPausedPresents} `
			+ `draw_pending=${runtime.isDrawPending || runtime.workbenchFaultState.faultSnapshot !== null ? 1 : 0} active_tick=${runtime.frameLoop.currentFrameState !== null ? 1 : 0}`
		);
		this.resetDebugCounters(currentTime);
	}
}

File: src/bmsx_cpp/render/presentation_state.cpp
#include "render/presentation_state.h"
#include "core/console.h"
#include "core/time.h"
#include "machine/runtime/runtime.h"
#include "render/shared/queues.h"
#include "render/test_pattern.h"
#include "render/vdp/view_snapshot.h"
#include <cstdio>
#include <cstdlib>

namespace bmsx {
namespace {

bool isPresentRateDebugEnabled() {
	static int cached = -1;
	if (cached < 0) {
		const char* env = std::getenv("BMSX_DEBUG_PRESENTRATE");
		cached = (env != nullptr && env[0] != '\0' && !(env[0] == '0' && env[1] == '\0')) ? 1 : 0;
	}
	return cached != 0;
}

} // namespace

void RenderPresentationState::recordHostFrame() {
	if (!isPresentRateDebugEnabled()) {
		return;
	}
	if (!m_debugPresentReportInitialized) {
		m_debugPresentReportInitialized = true;
		m_debugPresentReportAt = std::chrono::steady_clock::now();
	}
	m_debugPresentHostFrames += 1;
}

void RenderPresentationState::recordTickCompletion(bool visualCommitted, bool vdpFrameHeld) {
	if (!isPresentRateDebugEnabled()) {
		return;
	}
	m_debugPresentTickCompleted += 1;
	if (visualCommitted) {
		m_debugPresentTickCommitted += 1;
	} else {
		m_debugPresentTickDeferred += 1;
	}
	if (vdpFrameHeld) {
		m_debugPresentTickHeld += 1;
	}
}

void RenderPresentationState::recordPresentation(GameView::PresentationMode mode, bool commitFrame, bool paused) {
	if (!isPresentRateDebugEnabled()) {
		return;
	}
	if (paused) {
		m_debugPresentPausedPresents += 1;
		return;
	}
	if (mode == GameView::PresentationMode::Partial) {
		m_debugPresentPartialPresents += 1;
		return;
	}
	if (commitFrame) {
		m_debugPresentCommitPresents += 1;
		return;
	}
	m_debugPresentHoldPresents += 1;
}

void RenderPresentationState::flushDebugReport(const Runtime& runtime) {
	if (!isPresentRateDebugEnabled()) {
		return;
	}
	const auto now = std::chrono::steady_clock::now();
	if (!m_debugPresentReportInitialized) {
		m_debugPresentReportInitialized = true;
		m_debugPresentReportAt = now;
		return;
	}
	const double elapsedMs = to_ms(now - m_debugPresentReportAt);
	if (elapsedMs < 1000.0) {
		return;
	}
	const double scale = 1000.0 / elapsedMs;
	const double hostFps = static_cast<double>(m_debugPresentHostFrames) * scale;
	std::fprintf(
		stderr,
		"[BMSX] host_frames=%llu host_fps=%.2f ufps=%.2f tick_completed=%llu tick_committed=%llu tick_deferred=%llu tick_held=%llu present_partial=%llu present_commit=%llu present_hold=%llu present_paused=%llu draw_pending=%d active_tick=%d\n",
		static_cast<unsigned long long>(m_debugPresentHostFrames),
		hostFps,
		runtime.timing.ufps,
		static_cast<unsigned long long>(m_debugPresentTickCompleted),
		static_cast<unsigned long long>(m_debugPresentTickCommitted),
		static_cast<unsigned long long>(m_debugPresentTickDeferred),
		static_cast<unsigned long long>(m_debugPresentTickHeld),
		static_cast<unsigned long long>(m_debugPresentPartialPresents),
		static_cast<unsigned long long>(m_debugPresentCommitPresents),
		static_cast<unsigned long long>(m_debugPresentHoldPresents),
		static_cast<unsigned long long>(m_debugPresentPausedPresents),
		runtime.isDrawPending() ? 1 : 0,
		runtime.frameLoop.hasActiveTick(runtime) ? 1 : 0
	);
	m_debugPresentReportAt = now;
	m_debugPresentHostFrames = 0;
	m_debugPresentTickCompleted = 0;
	m_debugPresentTickCommitted = 0;
	m_debugPresentTickDeferred = 0;
	m_debugPresentTickHeld = 0;
	m_debugPresentPartialPresents = 0;
	m_debugPresentCommitPresents = 0;
	m_debugPresentHoldPresents = 0;
	m_debugPresentPausedPresents = 0;
}

void RenderPresentationState::clearPresentation() {
	m_pendingPresentation = false;
	m_presentationMode = GameView::PresentationMode::Completed;
	m_presentationCommitFrame = false;
}

void RenderPresentationState::reset() {
	clearPresentation();
	m_debugPresentReportInitialized = false;
	m_debugPresentHostFrames = 0;
	m_debugPresentTickCompleted = 0;
	m_debugPresentTickCommitted = 0;
	m_debugPresentTickDeferred = 0;
	m_debugPresentTickHeld = 0;
	m_debugPresentPartialPresents = 0;
	m_debugPresentCommitPresents = 0;
	m_debugPresentHoldPresents = 0;
	m_debugPresentPausedPresents = 0;
}

void RenderPresentationState::markPresentation(GameView::PresentationMode mode, bool commitFrame) {
	m_pendingPresentation = true;
	m_presentationMode = mode;
	m_presentationCommitFrame = commitFrame;
}

bool RenderPresentationState::consumePresentation(RenderPresentation& outPresentation) {
	if (!m_pendingPresentation) {
		return false;
	}
	outPresentation.mode = m_presentationMode;
	outPresentation.commitFrame = m_presentationCommitFrame;

	if (outPresentation.mode == GameView::PresentationMode::Completed && outPresentation.commitFrame) {
		RenderQueues::prepareCompletedRenderQueues();
	} else if (outPresentation.mode == GameView::PresentationMode::Completed) {
		RenderQueues::prepareHeldRenderQueues();
	} else {
		RenderQueues::preparePartialRenderQueues();
	}
	clearPresentation();
	return true;
}

void RenderPresentationState::syncAfterRuntimeUpdate(Runtime& runtime, i64 previousTickSequence) {
	if (runtime.frameScheduler.lastTickSequence != previousTickSequence) {
		markPresentation(GameView::PresentationMode::Completed, runtime.frameScheduler.lastTickVisualFrameCommitted);
	} else if (runtime.isDrawPending()) {
		markPresentation(GameView::PresentationMode::Partial, false);
	}
	while (runtime.frameScheduler.consumeTickCompletion(m_tickCompletionScratch)) {
		recordTickCompletion(m_tickCompletionScratch.visualCommitted, m_tickCompletionScratch.vdpFrameHeld);
	}
}

void RenderPresentationState::render(ConsoleCore& console, Runtime& runtime) {
	if (console.m_state != ConsoleState::Running && console.m_state != ConsoleState::Paused) {
		return;
	}

	const bool pausedPresent = console.m_state == ConsoleState::Paused;
	const bool runtimePresentPending = !pausedPresent && consumePresentation(m_presentationScratch);
	const bool shouldPresent = pausedPresent || runtimePresentPending;
	if (!shouldPresent) {
		return;
	}

	if (console.m_view) {
		const GameView::PresentationMode presentMode = pausedPresent
			? GameView::PresentationMode::Completed
			: m_presentationScratch.mode;
		const bool commitFrame = pausedPresent
			? false
			: m_presentationScratch.commitFrame;
		recordPresentation(presentMode, commitFrame, pausedPresent);
		if (pausedPresent) {
			RenderQueues::prepareHeldRenderQueues();
		}
		console.m_view->beginFrame();

		if (!console.m_rom_loaded) {
			renderTestPattern(*console.m_view, console.m_total_time);
		}

		commitVdpViewSnapshot(*console.m_view, runtime.machine().vdp());
		console.m_view->configurePresentation(presentMode, commitFrame);
		console.m_view->drawGame();
		console.m_view->endFrame();
	}

	flushDebugReport(runtime);
}

} // namespace bmsx
